File: frontend/src/types/master-data.ts
export interface Translation {
  [key: string]: string;
}

export enum MasterDataType {
  MAJOR = "MAJOR",
  COURSE_CODE = "COURSE_CODE",
  COURSE_LEVEL = "COURSE_LEVEL",
  DOCUMENT_CATEGORY = "DOCUMENT_CATEGORY",
}

export interface MasterData {
  id?: string;
  type: MasterDataType;
  code: string;
  translations: Translation;
  description?: string;
  createdAt?: Date;
  updatedAt?: Date;
  active: boolean;
  parentId?: string;
  fullUpdate?: boolean;
}

// Define the type for the hierarchy
type MasterDataHierarchy = {
  [key in MasterDataType]: MasterDataType[];
};

export const MASTER_DATA_HIERARCHY: MasterDataHierarchy = {
  [MasterDataType.MAJOR]: [],
  [MasterDataType.COURSE_CODE]: [MasterDataType.MAJOR],
  [MasterDataType.COURSE_LEVEL]: [],
  [MasterDataType.DOCUMENT_CATEGORY]: [],
};

File: document-interaction-service/src/main/java/com/dms/document/interaction/model/MasterData.java
package com.dms.document.interaction.model;

import com.dms.document.interaction.enums.MasterDataType;
import lombok.Data;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;

import java.time.Instant;

@Data
@Document(collection = "master_data")
@CompoundIndex(def = "{'type': 1, 'code': 1}", unique = true)
public class MasterData {
    @Id
    private String id;

    private MasterDataType type;
    private String code;
    private Translation translations;

    private String description;

    @Field("created_at")
    private Instant createdAt;

    @Field("updated_at")
    private Instant updatedAt;

    @Field("is_active")
    private boolean isActive;

    @Field("parent_id")
    private String parentId;
}